storage.go

```diff
@@ -109,64 +155,53 @@ func (s *Storage) listPrefix(ctx context.Context, prefix string, opt *pairStorag
 			return err
 		}
 
-		opt.ObjectFunc(o)
+		page.Data = append(page.Data, o)
 	}
-	return
-}
-func (s *Storage) metadata(ctx context.Context, opt *pairStorageMetadata) (meta info.StorageMeta, err error) {
-	meta = info.NewStorageMeta()
-	meta.Name = s.name
-	meta.WorkDir = s.workDir
-	return meta, nil
+
+	if header["X-List-Iter"] == codeIterEnd {
+		return IterateDone
+	}
+
+	input.marker = header["X-List-Iter"]
+	return nil
 }
-func (s *Storage) read(ctx context.Context, path string, opt *pairStorageRead) (rc io.ReadCloser, err error) {
+
+func (s *Storage) read(ctx context.Context, path string, w io.Writer, opt pairStorageRead) (n int64, err error) {
 	rp := s.getAbsPath(path)
 
-	var w *io.PipeWriter
-	rc, w = io.Pipe()
+	var pw *io.PipeWriter
+	var rc io.ReadCloser
+	rc, pw = io.Pipe()
 
 	go func() {
-		defer w.Close()
+		defer pw.Close()
 
 		_, err = s.bucket.Get(&upyun.GetObjectConfig{
 			Path:   rp,
-			Writer: w,
+			Writer: pw,
```

Comment:
Maybe we can pass the `w` directly here?

Reply:
If we pass the w directly, response body will be `Copy` into w. In this way, we cannot handle `IOCallback` as expected.

Reply:
We can implement a `iowrap.CallbackWriter`